pulser/waveforms.py

```diff
@@ -463,7 +483,7 @@ def samples(self) -> np.ndarray:
         Returns:
             numpy.ndarray: A numpy array with a value for each time step.
         """
-        return self._samples
+        return self._samples.copy()
```

Comment:
I'm thinking it might make sense to have all Waveforms defining a `self._samples` property instead of `self.samples`, and then have `self.samples` defined to return a copy of `self._samples` on the parent class. This way we can call `wf._samples` instead of `wf.samples` in `__getitem__` and avoid copying repeatedly. What you think?

Reply:
Working on it.

Reply:
I have doubts about the consequences and benefits. I must be missing something.

Currently, only `CustomWaveform` (before my commits) and `RampWaveform` (since my commits) are storing their samples, hence requiring a `copy()` before returning them.

Others (`BlackmanWaveform`, `CompositeWaveform` and `ConstantWaveform`) return arrays of samples computed on the fly, without a need for copy. Would you still add a `copy()` here?

Reply:
My idea would be for the ones that are computed on the fly, to only be once. Might be a good idea to use a [cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property) on `_samples`.

Reply:
I wonder if we're taking the right direction here.

I assume that, in the initial code, not saving samples was mostly a matter of saving memory.

I also assume that a call to the property `samples` will ALWAYS occur, sooner or later. Otherwise, why would we create a waveform anyway? :grin:

Hence, if we were to cache samples, we would ALWAYS end up with no memory gain.

Maybe we should just store samples in `__init__` most of the time (`ConstantWaveform` and `CompositeWaveform` possibly being exceptions). We would gain a lot in terms of code readability and speed of execution.

But I might be completely wrong along the way. Maybe I'm not taking parametrization into account.

Reply:
My concern is not really the memory gain, but rather the time wasted on multiple calls to `samples`. It's true that it always gets called once, in fact it gets called much more than once on average. The caching would simply be a way of only wasting time computing it once. That being said, it's very much the same thing as storing self._samples in __init__. The only difference is that we can mark `self._samples` with `@abstractmethod` on the parent class, so it enforces that all child classes have `self._samples` defined.

To be clear, `@cached_property` will run the first time it's called and store the result. All ensuing calls will essentially lookup the stored value. I feel like it is in line with what you're suggesting, with some added benefits.

Reply:
OK, I'll be adding `@cached_property` then.

BTW, since it always gets called once (at least), maybe we can improve/simplify some of the code I wrote for `__getitem__`.

For instance, inside `CompositeWaveform`: what about computing/caching samples once and for all, and then simply index them, rather than using the loops we've been working on (so hard)?

Reply:
Yes, I think that's sensible. I'm sorry that it will make you get rid of the loops, though...

Reply:
We'll miss the loops but we'll forever remember and cherish the nice work we achieved on them. 😁

If we are to create `samples` in the parent class as `return self._samples.copy()`:
* That would also apply to `ConstantWaveform`, in which case we should cache the constant samples instead of always returning a `np.full()` built on the fly?
* All `__getitem__` could be simplified as mere `return self._samples[index_or_slice]` in the parent class?
* Except maybe with a few optimizations, like:
  * `ConstantWaveform`: Simply returning the constant value when `index_or_slice` is an `int` instead of indexing `self._samples`?
  * `BlackmanWaveform` and `RampWaveform`: Simply returning the corresponding stored values when we index the first or last sample?

Reply:
Yes to all!